### PR TITLE
Systematically check that JSON-LD types == Python types.

### PR DESCRIPTION
**Pull request by marcenacp**
- Add more documentation.
- Isolate logic to cast types in dedicated functions.
- Check types (which allowed me to find some issues in the code).
- Add unit tests.

### CODE REVIEW

File: python/mlcroissant/README.md [hunk 1]
```diff
@@ -127,6 +127,19 @@ class RecordSet(Node):
     )
 ```
 
+**Example 3**: specify a version:
```
Comment (ccl-core) [inline python/mlcroissant/README.md:130]: NIT I would probably say something about the fact that if none given it is V1.
And that, in some cases, it is important to select another version, like if a property is only available for a specific version, like below

File: python/mlcroissant/mlcroissant/_src/core/versions.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+"""Versions module."""
+
+from typing import Any
+
+from mlcroissant._src.core.issues import ErrorException
+from mlcroissant._src.core.issues import WarningException
+
+
+def cast_version(version: Any) -> str | None:
+    """Casts the given version and returns a normalized string version.
```
Comment (ccl-core) [inline python/mlcroissant/mlcroissant/_src/core/versions.py:10]: NIT: dataset version (to not confuse this with specs version?)

File: python/mlcroissant/mlcroissant/_src/structure_graph/base_node.py [hunk 3]
```diff
@@ -53,13 +62,27 @@ class Node:
     ctx: Context = dataclasses.field(default_factory=Context)
     id: str = dataclasses.field(default_factory=generate_uuid)
     name: str | None = None
-    parents: list[Node] = dataclasses.field(default_factory=list)
+    parents: list["Node"] = dataclasses.field(default_factory=list)
```
Comment (ccl-core) [inline python/mlcroissant/mlcroissant/_src/structure_graph/base_node.py:65]: Out of curiosity, why is this needed?

**Review by ccl-core — APPROVED** [review]

File: python/mlcroissant/README.md (hunk 1 above)
Comment (marcenacp) [reply python/mlcroissant/README.md:130]: Done.

By default, all versions are supported, not just v1.

File: python/mlcroissant/README.md [hunk 4]
```diff
@@ -127,6 +127,19 @@ class RecordSet(Node):
...
+
+```python
+@mlc_dataclasses.dataclass
+class Field(Node):
+    is_enumeration: bool | None = mlc_dataclasses.jsonld_field(
+        default=None,
+        input_types=[SDO.Boolean],
+        url=constants.ML_COMMONS_IS_ENUMERATION,
+        versions=[CroissantVersion.V_0_8],   # `is_enumeration` will only be valid for v0.8, not v1.0
```
[the remark this answers is not in the document]
Comment (marcenacp) [reply python/mlcroissant/README.md:139]: Done.

File: python/mlcroissant/mlcroissant/_src/core/versions.py (hunk 2 above)
Comment (marcenacp) [reply python/mlcroissant/mlcroissant/_src/core/versions.py:10]: Done.

File: python/mlcroissant/mlcroissant/_src/structure_graph/base_node.py (hunk 3 above)
Comment (marcenacp) [reply python/mlcroissant/mlcroissant/_src/structure_graph/base_node.py:65]: I removed `from __future__ import annotations` in order for [`dataclasses.field`](https://docs.python.org/3/library/dataclasses.html#dataclasses.Field)'s types to be populated with types and not strings. Without this annotation, we need to use strings.

**Comment by marcenacp** [comment]
Thanks for the review, @ccl-core!

File: python/mlcroissant/README.md (hunk 1 above)
Comment (marcenacp) [reply python/mlcroissant/README.md]: PS: I also added documentation in python/mlcroissant/mlcroissant/_src/core/dataclasses.py (Metadata's docstring).
